src/lib/autoinstall/ask/runner.rb

```diff
@@ -0,0 +1,239 @@
+    # the response implies:
+    #
+    # * Updating the profile if an AutoYaST path was given (elements `path` and
+    #   `pathlist` from the profile).
+    # * Writing the answer to the file, if a `file` element was specified.
+    # * Running the given script (`script` element from the profile).
+    #
+    # @example Ask questions from the profile during the 1st stage
+    #   Yast::Profile.ReadXML("autoinst.xml")
+    #   runner = Runner.new(Yast::Profile.current, stage: Stage::INITIIAL)
+    #   runner.run
+    class Runner
+      include Yast::Logger
+
+      # @return [Hash] AutoYaST profile
+      attr_reader :profile
+      # @return [Stage] Stage to run; only the dialogs/questions for the given
+      #   stage are shown.
+      attr_reader :stage
+
+      # @param profile [Hash] AutoYaST profile
+      # @param stage [Stage] Consider dialogs/questions from the given stage
+      def initialize(profile, stage:)
+        @profile = profile
+        @indexes = []
+        @stage = stage
+      end
+
+      # Runs the dialogs and processes the user input
+      #
+      def run
+        current_dialog = go_next
+        loop do
+          break if current_dialog.nil?
+
+          result = run_dialog(current_dialog)
+          if result == :back
+            current_dialog = go_back
+          elsif result == :next
+            results = current_dialog.questions.map { |q| process_question(q) }
+            current_dialog = go_next unless results.any? { |r| r == :repeat }
+          end
+        end
+      end
+
+    private
+
+      # Returns the AskListSection from the profile
+      #
+      # @return [AskListSection]j
+      def ask_list
+        return @ask_list if @ask_list
+
+        section = profile.fetch_as_hash("general").fetch_as_array("ask-list")
+        @ask_list = Y2Autoinstall::AutoinstProfile::AskListSection.new_from_hashes(section)
+      end
+
+      # Run the given dialog
+      #
+      # @param dialog [Y2Autoinstall::Ask::Dialog] Ask dialog specification
+      # @return [Symbol] :next or :back
+      def run_dialog(dialog)
+        ask_dialog = Y2Autoinstall::Widgets::Ask::Dialog.new(
+          dialog, disable_back_button: first?
+        )
+        ask_dialog.run
+      end
+
+      DIALOG_FILE = "/tmp/next_dialog".freeze
+      MAX_DIALOG_FILE_SIZE = 1024
+      private_constant :MAX_DIALOG_FILE_SIZE
+
+      # Finds the ID of the next dialog
+      #
+      # It checks for a regular file located at {DIALOG_FILE} which contains
+      # the number of the following dialog. If it does not exist, it goes with
+      # the dialog after the current one.
+      #
+      # For security reasons, it checks whether the size of the file makes
+      # sense for this use case, to not load a big chunk of data into memory by
+      # mistake.
+      #
+      # @return [Integer,nil] Next dialog ID
+      def find_next_dialog_index
+        return unless File.file?(DIALOG_FILE) && File.size(DIALOG_FILE) <= MAX_DIALOG_FILE_SIZE
```

Comment:
I would log if it failed for any reason ( I mean not if file is not there, but if file is not regular file or size is too big ).